src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs

```diff
@@ -368,11 +367,43 @@ protected virtual SamlConditions CreateConditions(SecurityTokenDescriptor tokenD
             else if (SetDefaultTimesOnTokenCreation)
                 conditions.NotOnOrAfter = DateTime.UtcNow + TimeSpan.FromMinutes(TokenLifetimeInMinutes);
 
-            if (!string.IsNullOrEmpty(tokenDescriptor.Audience))
-                conditions.Conditions.Add(new SamlAudienceRestrictionCondition(new Uri(tokenDescriptor.Audience)));
+            if (!tokenDescriptor.Audiences.IsNullOrEmpty())
+                if (!tokenDescriptor.Audience.IsNullOrEmpty())
+                    conditions.Conditions.Add(CreateAudienceRestrictionCondition(tokenDescriptor.Audience, tokenDescriptor.Audiences));
+                else
+                    conditions.Conditions.Add(CreateAudienceRestrictionCondition(tokenDescriptor.Audiences));
+
+            else if(!tokenDescriptor.Audience.IsNullOrEmpty())
+                conditions.Conditions.Add(CreateAudienceRestrictionCondition(tokenDescriptor.Audience));
 
             return conditions;
         }
+        private static SamlAudienceRestrictionCondition CreateAudienceRestrictionCondition(string audience)
+        {
+            SamlAudienceRestrictionCondition audRestrictionCondition = new ();
+            audRestrictionCondition.Audiences.Add(new Uri(audience));
+            return audRestrictionCondition;
+        }
+
+        private static SamlAudienceRestrictionCondition CreateAudienceRestrictionCondition(IList<string> audiences)
+        {
+            SamlAudienceRestrictionCondition audRestrictionCondition = new();
+            foreach (var audience in audiences)
+                audRestrictionCondition.Audiences.Add(new Uri(audience));
+
+            return audRestrictionCondition;
+        }
+
+        private static SamlCondition CreateAudienceRestrictionCondition(string audience, IList<string> audiences)
+        {
+            SamlAudienceRestrictionCondition audRestrictionCondition = new();
```

Comment:
Could you use this constructor?
 public SamlAudienceRestrictionCondition(IEnumerable<Uri> audiences)

Reply:
Not without making an extra new List/Enumerable because that constructor takes IEnumerable <Uri> whereas the list we have is IList<string>.